Optimizing uint operations (architecture independent)

tomusdrw opened the pull request:
Related to #490 

Starting from:
```
test u128_mul      ... bench:   2,305,342 ns/iter (+/- 107,978)
test u256_add      ... bench:     135,920 ns/iter (+/- 9,476)
test u256_full_mul ... bench:  15,597,671 ns/iter (+/- 593,695)
test u256_mul      ... bench:   5,241,255 ns/iter (+/- 316,925)
test u256_sub      ... bench:     545,754 ns/iter (+/- 44,429)
test u512_add      ... bench:     191,064 ns/iter (+/- 10,624)
test u512_sub      ... bench:   1,627,907 ns/iter (+/- 82,830)
```
Current:
```
test u128_mul      ... bench:      45,772 ns/iter (+/- 1,508)
test u256_add      ... bench:      30,052 ns/iter (+/- 2,552)
test u256_full_mul ... bench:   3,005,477 ns/iter (+/- 98,643)
test u256_mul      ... bench:     101,664 ns/iter (+/- 7,941)
test u256_sub      ... bench:      32,904 ns/iter (+/- 1,382)
test u512_add      ... bench:      78,783 ns/iter (+/- 3,864)
test u512_sub      ... bench:      59,469 ns/iter (+/- 2,340)
```


vs ASM
```
test u128_mul      ... bench:      52,687 ns/iter (+/- 2,566)
test u256_add      ... bench:      38,811 ns/iter (+/- 2,439)
test u256_full_mul ... bench:   2,482,358 ns/iter (+/- 141,787)
test u256_mul      ... bench:     124,951 ns/iter (+/- 6,652)
test u256_sub      ... bench:      38,449 ns/iter (+/- 1,872)
test u512_add      ... bench:      96,885 ns/iter (+/- 6,237)
test u512_sub      ... bench:      96,400 ns/iter (+/- 5,347)
```

NikVolf commented on util/src/lib.rs:
```diff
@@ -167,3 +167,59 @@ pub use io::*;
 pub use kvdb::*;
 
+#[cfg(test)]
+mod tests {
```
why move here?

tomusdrw replied:
Those tests were not compiling in `bigint` crate - they are for `H256` which is implemented in `util/src/hash.rs`

NikVolf commented:
```
test u512_add      ... bench:      79,473 ns/iter (+/- 2,289)
test u512_sub      ... bench:         461 ns/iter (+/- 14)
```

why so much difference?

debris commented:
Maybe benchmark code was incorrectly inlined?

tomusdrw commented:
Yeah, it must have been some heavy optimization by compiler (the subtraction was done only for most significant u64). I've modified the bench and updated results.

NikVolf commented:
looks good to me, though with next pr we should move fixed hash integers to bigint and return tests

gavofyork commented:
will hold off until after 1.0.

gavofyork commented:
conflicts need sorting.

petevine commented:
I've updated my benchmark results in https://github.com/ethcore/parity/issues/596#issuecomment-192630065, well done!
